fix(eslint): Use @typescript-eslint extension rule for no-unused-expressions

ryanb-jobber opened the pull request:
## Motivations

The default eslint rule `no-unused-expressions` flags conditionally chained functions as an error. 

e.g. 
`object?.function()` results in a `no-unused-expressions` error.

To enable this, `no-unused-expressions` has been set to off. 

The typescript rule `@typescript-eslint/no-unused-expressions` allows this syntax, while still keeping the originally intended lint rule intact.

## Changes

<!-- https://keepachangelog.com/en/1.0.0/ -->

### Changed

- Added rule `@typescript-eslint/no-unused-expressions`
- Added rule `@typescript-eslint/no-empty-function` and fixed violations in tests

## Testing

Using the syntax `object?.function()` should not be flagged with the linter. Other unused expressions should be flagged.

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

ryanb-jobber commented on packages/eslint-config/package.json:
```diff
@@ -8,7 +8,7 @@
   "dependencies": {
     "@rushstack/eslint-patch": "^1.0.2",
-    "@typescript-eslint/eslint-plugin": "^2.1.0",
+    "@typescript-eslint/eslint-plugin": "^2.7.0",
```
Version bump required to use the no-unused-expressions rule.

ryanb-jobber commented on packages/eslint-config/.eslintrc.js:
```diff
@@ -71,6 +71,9 @@ module.exports = {
     "@typescript-eslint/consistent-type-definitions": ["error", "interface"],
+    "@typescript-eslint/no-unused-expressions": ["error", { allowShortCircuit: true }],
+    "no-empty-function": "off",
+    "@typescript-eslint/no-empty-function": ["error", { "allow": ["arrowFunctions"] }],
```
I got several empty function errors related to arrow functions in tests after updating the eslint-plugin version, so I suspect the default inspection rules changed.

I added configuration here to allow empty arrow functions.

seanhealy replied:
@ryanb-jobber do you recall where the errors were coming from? If there aren't many it might be nice to fix them as part of this…

ryanb-jobber replied:
@seanhealy 

They were empty click handlers in 3 different test suites. I replaced them all with mock functions and that seemed to do the trick, so I removed the arrow function exception 😄

chris-at-jobber commented:
@ryanb-jobber @seanhealy would this still be useful? Looks like it was waiting on review

ryanb-jobber commented:
@chris-at-jobber @seanhealy I think so! This was done in jobber-mobile, this change was to make Atlantis equivalent in terms of lint rules.

joelwarrington-jobber reviewed (changes requested):
I believe this PR a bit stale - but think it's worth getting this in, do you mind resolving conflicts and then can review again?

ryanb-jobber commented:
Addressed merge issues and updated PR description to reflect the current state (it looks like the original eslint rule was turned off at some point to get around this)

joelwarrington-jobber reviewed (approved):
LGTM
